pgmpy/estimators/TreeAugmentedNaiveBayesSearch.py

```diff
@@ -0,0 +1,103 @@
+        """
+        Search class for learning tree-augmented naive bayes (TAN) graph structure with a given set of variables.
+
+        Parameters
+        ----------
+        data: pandas DataFrame object
+            dataframe object where each column represents one variable.
+
+        class_node: node
+            Dependent variable of the model (i.e. the class label to predict)
+
+        root_node: node (optional)
+            The root node of the tree structure over the independent variables.  If not specified, then
+            an arbitrary independent variable is selected as the root.
+
+        state_names: dict (optional)
+            A dict indicating, for each variable, the discrete set of states (or values)
+            that the variable can take. If unspecified, the observed values in the data set
+            are taken to be the only possible states.
+
+        complete_samples_only: bool (optional, default `True`)
+            Specifies how to deal with missing data, if present. If set to `True` all rows
+            that contain `np.Nan` somewhere are ignored. If `False` then, for each variable,
+            every row where neither the variable nor its parents are `np.NaN` is used.
+            This sets the behavior of the `state_count`-method.
+
+        References
+        ----------
+        Friedman N, Geiger D and Goldszmidt M (1997). Bayesian network classifiers. Machine Learning 29: 131–163
+
+        """
+        self.class_node = class_node
+        self.root_node = root_node
+
+        super(TreeAugmentedNaiveBayesSearch, self).__init__(data, **kwargs)
+
+    def estimate(self):
+        """
+        Estimate the `DAG` structure that fits best to the given data set without parametrization.
+
+        Returns
+        -------
+        model: `DAG` instance
+            A tree augmented naive bayes `DAG`.
+
+        Examples
+        --------
+        >>> import numpy as np
+        >>> import pandas as pd
+        >>> import networkx as nx
+        >>> import matplotlib.pyplot as plt
+        >>> from pgmpy.estimators import ExhaustiveSearch
+        >>> values = pd.DataFrame(np.random.randint(low=0, high=2, size=(1000, 5)), columns=['A', 'B', 'C', 'D', 'E'])
+        >>> class_node = 'A'
+        >>> est = TreeAugmentedNaiveBayesSearch()
+        >>> model = est.estimate()
+        >>> nx.draw_circular(model, with_labels=True, arrowsize=20, arrowstyle='fancy', alpha=0.3)
+        >>> plt.show()
+
+        """
+        if self.class_node not in self.data.columns:
+            raise ValueError("class node must exist in data")
+
+        if self.root_node is not None and self.root_node not in self.data.columns:
+            raise ValueError("root node must exist in data")
+
+        # construct maximum spanning tree
+        G = nx.Graph()
+        df_features = self.data.loc[:, self.data.columns != self.class_node]
+        total_cols = len(df_features.columns)
```

Comment:
If I am understanding this code block correctly, you are trying to compute the mutual info score between all possible combination of variables, creating an edge between them and using the mutual info as the edge weight. I think this can be simplified by computing an array of edge_weights and then creating the graph from that. Something like:
```
weights = pd.DataFrame(0, index=df_features.columns, columns=df_features.columns)
for u_index, u, v_index, v in combinations(enumerate(df_features.columns), 2):
   weights.iat[u_index, v_index] = mutual_info_score(df_features.iloc[:, u_index], df_features.iloc[:, v_index])
G = nx.from_pandas_adjacency(weights, create_using=nx.Graph)
```
I think the main benefit would be that this would be much faster because networkx's `add_node` and `add_edge` can be super slow. Also, I think this is more readable.

Reply:
You can also add a nice `tqdm` progress bar and make the computation parallel at this step.

Reply:
Sorry, I'm not familiar with multi-threading in Python.  Do you mean `joblib` and the step to parallelize is:

`weights.iat[u_index, v_index] = mutual_info_score(df_features.iloc[:, u_index], df_features.iloc[:, v_index])`

I assume Pandas is thread-safe, and/or concurrent writes wouldn't cause any race condition or deadlocks to the underlying data structures backing the dataframe.

Reply:
Yes, I was suggesting using `joblib`. We can run the `for` loop in parallel as there is no order requirement in this case. You can refer to: https://joblib.readthedocs.io/en/latest/parallel.html for parallelizing for loops using joblib. We will also need an argument `n_jobs=-1` in the method for the user to specify the number of threads they want to use. But this isn't a priority change, so can be done later as well.